### PR TITLE
Forward / backward NJT support for several activation functions

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141604
* #141506
* #141392
* #140161
* __->__ #140736
* #141500

Several activation functions were unimplemented due to missing `pointwise` tags. This PR adds them and corresponding backwards implementations.

### CODE REVIEW

File: torch/nested/_internal/ops.py [hunk 1]
```diff
@@ -2212,6 +2212,41 @@ def new_empty_default(func, *args, **kwargs):
...
+        torch.ops.aten.hardsigmoid_backward.default,
+        torch.ops.aten.hardtanh_backward.default,
+        torch.ops.aten.softplus_backward.default,
+        torch.ops.aten.softshrink_backward.default,
+    ],
+    "self: jt_all, ...",
+)
+def activation_backward(func, *args, **kwargs):
+    # first NJT arg is expected to be grad_output
+    grad_output = next(arg for arg in args if isinstance(arg, NestedTensor))
```
Comment (Skylion007) [inline torch/nested/_internal/ops.py:2228]: Add an. assert that arg is not empty?

**Review by Skylion007 — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -5031,6 +5031,7 @@
 
 - func: relu6(Tensor self) -> Tensor
   python_module: nn
+  tags: pointwise
 
 - func: relu6_(Tensor(a!) self) -> Tensor(a!)
```
Comment (Skylion007) [inline aten/src/ATen/native/native_functions.yaml:5036]: The inplace versions should also be marked as pointwise, right?

File: torch/nested/_internal/ops.py (hunk 1 above)
Comment (jbschlosser) [reply torch/nested/_internal/ops.py:2228]: `check_schema()` handles this part :)

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (Skylion007) [reply aten/src/ATen/native/native_functions.yaml:5036]: Ah, but lots of other inplace ops are already tagged as pointwise unfortunately.

File: torch/nested/_internal/ops.py (hunk 1 above)
Comment (Skylion007) [reply torch/nested/_internal/ops.py:2228]: But couldn't none of the args be NestedTensors?

**Review by cpuhrsch — APPROVED** [review]

File: torch/nested/_internal/ops.py (hunk 1 above)
Comment (soulitzer) [reply torch/nested/_internal/ops.py:2228]: We wouldn't have reached this torch dispatch if none are NestedTensors

Comment (jbschlosser) [reply torch/nested/_internal/ops.py:2228]: `check_schema()` will validate that `self` is an NJT so we're guaranteed at least one, plus we're inside of `NestedTensor.__torch_dispatch__()` so we had to get there somehow from a `NestedTensor` instance

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml:5036]: hm true, I think I'll open a follow-up to handle this properly

**Review by Skylion007 — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml:5036]: oh, this actually already works today:
```python
def test_inplace_unary_op_returns_same_instance(self, device, dtype):
    nt = random_nt_from_dims(
        [3, None, 5],
        device=device,
        dtype=dtype,
        layout=torch.jagged,
    )
    out = nt.relu_()
    # passes!
    self.assertIs(out, nt)
```

What I believe happens is that the ADInplaceOrView key handles aliasing.

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but its tests are failing in trunk' -c nosignal

test_nestedtensor.py::TestNestedTensorOpInfoCUDA::test_compile_backward_nn_functional_softshrink_cuda_float32 [GH job link](https://github.com/pytorch/pytorch/actions/runs/11980859033/job/33407057580) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/af70f5e04c69839a1a0e08942254c170dc4c3d61)

**Pull request closed without merging**